src/format-promoted-commits.ts

```diff
@@ -0,0 +1,168 @@
+        );
+      }
+
+      const { promotionSet, dockerImageRepository } =
+        promotionSetWithDockerImage;
+      const promotionSetJSON = JSON.stringify(promotionSet);
+
+      let promotionSetWithApps =
+        organizedPromotionsByPromotionSetJSON.get(promotionSetJSON);
+      if (!promotionSetWithApps) {
+        promotionSetWithApps = { promotionSet, apps: [] };
+        organizedPromotionsByPromotionSetJSON.set(
+          promotionSetJSON,
+          promotionSetWithApps,
+        );
+      }
+
+      promotionSetWithApps.apps.push({ appDirectory, dockerImageRepository });
+    }
+  }
+  return organizedPromotionsByTargetEnvironment;
+}
+
+export function formatPromotedCommits(
+  promotionsByFileThenEnvironment: Map<string, PromotionsByTargetEnvironment>,
+): string {
+  const organizedPromotionsByTargetEnvironment =
+    reorganizePromotionInfoForMessage(promotionsByFileThenEnvironment);
+  return [...organizedPromotionsByTargetEnvironment.entries()]
+    .sort(([a], [b]) => a.localeCompare(b))
+    .map(([targetEnvironment, organizedPromotionsByPromotionSetJSON]) => {
+      const environmentHeader = `### Promoting to ${targetEnvironment}\n`;
+      const forEnvironment = [
+        ...organizedPromotionsByPromotionSetJSON.entries(),
+      ]
+        // We sort by the JSON value just to keep things consistent when we
+        // refresh.
+        .sort(([a], [b]) => a.localeCompare(b))
+        .map((entry) => {
+          const { promotionSet, apps } = entry[1];
+          const {
+            trimmedRepoURL,
+            gitConfigPromotionInfo,
+            dockerImagePromotionInfo,
+            links,
+          } = promotionSet;
+          const text = [
+            `Apps:\n${apps
+              .sort((a, b) => a.appDirectory.localeCompare(b.appDirectory))
+              .map(({ appDirectory, dockerImageRepository }) =>
+                dockerImageRepository &&
+                dockerImageRepository !== basename(appDirectory)
+                  ? `- ${appDirectory} (image \`${dockerImageRepository}\`)\n`
+                  : `- ${appDirectory}\n`,
+              )
+              .join('')}\n`,
+          ];
+          if (links.length) {
+            text.push(`Links:\n`);
+            for (const link of links) {
+              if (link.url) {
+                text.push(`- [${link.text}](${link.url})\n`);
+              } else {
+                text.push(`- ${link.text}\n`);
+              }
+            }
+            text.push(`\n`);
+          }
+          let alreadyMentionedGitNoCommits = false;
+          if (
+            dockerImagePromotionInfo &&
+            dockerImagePromotionInfo.type !== 'no-change'
+          ) {
+            let maybeGitConfigNoCommits = '';
+            if (gitConfigPromotionInfo.type === 'no-commits') {
+              alreadyMentionedGitNoCommits = true;
+              maybeGitConfigNoCommits =
+                ' (and the git ref for the Helm chart made a no-op change to match)';
```

Comment:
'(with a matching no-op Helm chart update)' maybe?  in practice, does the user even care about this?  What if we just left it off?

Reply:
gonna try "(matching Helm chart update is a no-op)" for now... I do kinda feel like answering the question "why did I see a change to the git config that isn't being mentioned" sort of matters but maybe it can be even less words (like an asterisk or emoji that's defined at the bottom?)